File: python/keystore/keystore_crypto.py
"""Keystore encryption: scrypt + AES-256-GCM (compatible with TypeScript payload)."""
import base64
import hashlib
import os
from typing import Any

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

KEY_LEN = 32
SALT_LEN = 16
IV_LEN = 12
TAG_LEN = 16
SCRYPT_N = 16384
SCRYPT_R = 8
SCRYPT_P = 1


def _derive_key(password: str, salt: bytes) -> bytes:
    return hashlib.scrypt(
        password.encode("utf-8"),
        salt=salt,
        n=SCRYPT_N,
        r=SCRYPT_R,
        p=SCRYPT_P,
        dklen=KEY_LEN,
    )


def encrypt(plaintext: str, password: str) -> dict[str, Any]:
    """Encrypt plaintext with password. Returns payload (version, salt, iv, tag, data) matching TS."""
    salt = os.urandom(SALT_LEN)
    iv = os.urandom(IV_LEN)
    key = _derive_key(password, salt)
    aes = AESGCM(key)
    plainbytes = plaintext.encode("utf-8")
    ct_with_tag = aes.encrypt(iv, plainbytes, None)
    ciphertext_only = ct_with_tag[:-TAG_LEN]
    tag = ct_with_tag[-TAG_LEN:]
    return {
        "version": 1,
        "salt": salt.hex(),
        "iv": iv.hex(),
        "tag": tag.hex(),
        "data": base64.b64encode(ciphertext_only).decode("ascii"),
    }


def decrypt(payload: dict[str, Any], password: str) -> str:
    """Decrypt payload with password. Returns plaintext string."""
    salt = bytes.fromhex(payload["salt"])
    iv = bytes.fromhex(payload["iv"])
    tag = bytes.fromhex(payload["tag"])
    if len(salt) != SALT_LEN:
        raise ValueError(f"Invalid salt length: expected {SALT_LEN}, got {len(salt)}")
    if len(iv) != IV_LEN:
        raise ValueError(f"Invalid iv length: expected {IV_LEN}, got {len(iv)}")
    if len(tag) != TAG_LEN:
        raise ValueError(f"Invalid tag length: expected {TAG_LEN}, got {len(tag)}")
    # Always decode data as base64 (matches TypeScript encrypt which always outputs base64)
    ciphertext_only = base64.b64decode(payload["data"])
    ciphertext = ciphertext_only + tag
    key = _derive_key(password, salt)
    aes = AESGCM(key)
    plainbytes = aes.decrypt(iv, ciphertext, None)
    return plainbytes.decode("utf-8")


def is_encrypted_payload(obj: Any) -> bool:
    """Return True if obj looks like an encrypted payload (version, salt, iv, tag, data)."""
    if not isinstance(obj, dict):
        return False
    return (
        obj.get("version") == 1
        and isinstance(obj.get("salt"), str)
        and isinstance(obj.get("iv"), str)
        and isinstance(obj.get("tag"), str)
        and isinstance(obj.get("data"), str)
    )

File: typescript/src/keystore/keystore_crypto.ts
import * as crypto from 'crypto';

const ALGORITHM = 'aes-256-gcm';
const KEY_LEN = 32;
const SALT_LEN = 16;
const IV_LEN = 12;
const TAG_LEN = 16;
const SCRYPT_N = 16384;
const SCRYPT_R = 8;
const SCRYPT_P = 1;

export interface EncryptedPayload {
  version: number;
  salt: string;
  iv: string;
  tag: string;
  data: string;
}

/**
 * Derive key from password using scrypt
 */
function deriveKey(password: string, salt: Buffer): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    crypto.scrypt(password, salt, KEY_LEN, { N: SCRYPT_N, r: SCRYPT_R, p: SCRYPT_P }, (err, key) => {
      if (err) reject(err);
      else resolve(key);
    });
  });
}

/**
 * Encrypt plaintext with password. Returns payload with salt, iv, tag, data (all hex/base64).
 */
export async function encrypt(plaintext: string, password: string): Promise<EncryptedPayload> {
  const salt = crypto.randomBytes(SALT_LEN);
  const iv = crypto.randomBytes(IV_LEN);
  const key = await deriveKey(password, salt);
  const cipher = crypto.createCipheriv(ALGORITHM, key, iv);
  const enc = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return {
    version: 1,
    salt: salt.toString('hex'),
    iv: iv.toString('hex'),
    tag: tag.toString('hex'),
    data: enc.toString('base64')
  };
}

/**
 * Decrypt payload with password. Returns plaintext string.
 */
export async function decrypt(payload: EncryptedPayload, password: string): Promise<string> {
  const key = await deriveKey(password, Buffer.from(payload.salt, 'hex'));
  const decipher = crypto.createDecipheriv(
    ALGORITHM,
    key,
    Buffer.from(payload.iv, 'hex')
  );
  decipher.setAuthTag(Buffer.from(payload.tag, 'hex'));
  return decipher.update(payload.data, 'base64', 'utf8') + decipher.final('utf8');
}

/**
 * Check if a JSON object looks like an encrypted payload (has version, salt, iv, tag, data).
 */
export function isEncryptedPayload(obj: unknown): obj is EncryptedPayload {
  const o = obj as Record<string, unknown>;
  return (
    o &&
    typeof o === 'object' &&
    o.version === 1 &&
    typeof o.salt === 'string' &&
    typeof o.iv === 'string' &&
    typeof o.tag === 'string' &&
    typeof o.data === 'string'
  );
}
